PropertyGraph: move non-typical access functions to another class

tylershunt opened the pull request:
Sometimes we have to to do messy things to PropertyGraphs especially
when partitioning them. In an effort to keep those interfaces from
confusing direct users of PropertyGraph, this PR factors them out into
a friend class that I'm calling a "Retractor". The retractor lets
other codes that knows what it's doing have controlled access to
PropertyGraph internals.

This was motivated by the smaller change which added methods to
deallocate topology information. (Required to save memory during
repartitioning).

Signed-off-by: Tyler Hunt <thunt@katanagraph.com>

roshandathathri commented on libgalois/include/galois/graphs/LC_CSR_Graph.h:
```diff
@@ -760,6 +760,65 @@ class LC_CSR_Graph :
+            std::enable_if_t<!std::is_same<E, void>::value, int>* = nullptr>
+  void constructFrom(
+      uint32_t numNodes, uint64_t numEdges,
+      galois::LargeArray<uint64_t>& prefix_sum,
```
Use a move parameter to make it clear to the user that the array will be moved.

roshandathathri commented on lonestar/analytics/cpu/bipart/Helper.h:
```diff
@@ -0,0 +1,74 @@
+#include "Bipart.h"
+
+template <typename T>
+uint64_t ParallelPrefixSum(galois::LargeArray<T>& prefix_sum) {
```
This should be in libgalois as this is not specific to bipart.

Moreover, there is an existing parallel prefix sum in libpangolin/include/pangolin/scan.h. Move that to libgalois and merge this function with that.

roshandathathri commented on lonestar/analytics/cpu/bipart/Helper.h:
```diff
@@ -0,0 +1,74 @@
+      ++block_size;
+    }
+    uint32_t start = tid * block_size;
+    uint32_t end   = (tid + 1) * block_size;
```
Better to use block_range in gstl.h instead of explicitly creating blocks.

roshandathathri commented on lonestar/analytics/cpu/bipart/Helper.h:
```diff
@@ -0,0 +1,74 @@
+      prefix_sum[idx] += prefix_sum[idx - 1];
+    }
+
+    interm_sums[tid] = prefix_sum[end - 1];
```
interm_sums could have cache conflicts because they might be on the same cache line. It's not performance critical as it is just one write per thread but using thread-local variables instead might be a better design.

roshandathathri commented on lonestar/analytics/cpu/bipart/Helper.h:
```diff
@@ -0,0 +1,74 @@
+    uint32_t end   = (tid + 1) * block_size;
+    if (end > size) {
+      end = size;
+    }
```
Better to use block_range in gstl.h instead of explicitly creating blocks.

roshandathathri commented on lonestar/analytics/cpu/bipart/README.md:
```diff
@@ -23,10 +23,15 @@ BUILD
+
+        RUN-- ------------------------------------------------------------------------------
+
+    To run on machine with a k value of 4,
```
Formatting seems to be messed up?

roshandathathri commented on lonestar/analytics/cpu/bipart/Bipart.cpp:
More descriptive names/strings for the user please.

gurbinder533 replied:
@uditagar can you please provide a more descriptive name for these policies? Thank you.

roshandathathri commented on lonestar/analytics/cpu/bipart/Bipart.cpp:
mtx_input

Also, if this is supported, it's better to move this to libbench in lonestar so that other applications can also benefit from this.

gurbinder533 replied:
It is not being used, so removing this along with other unused arguments.

roshandathathri commented on lonestar/analytics/cpu/bipart/Bipart.cpp:
This seems similar to an existing option in lonestar libbench or libdistbench. Please use a different name. Preferably something that exists in libdistbench to store partitions.

roshandathathri commented on lonestar/analytics/cpu/bipart/Bipart.cpp:
What is the definition of such a graph? This should be included in the description. 

Symmetric graph, hyper graph, etc. have well defined semantics. I don't know what hMetisGraph means.

roshandathathri commented on lonestar/analytics/cpu/bipart/Bipart.cpp:
Same comment as outfile

nicelhc13 replied:
How about removing this option? We only need `output_file_name`  option like statFile option of the general lonestar. And we can rename the `output_file_name` to `outfile` as Roshan suggested.

roshandathathri commented on lonestar/analytics/cpu/bipart/Bipart.cpp:
Isn't this unused? How is this not a compiler error!

gurbinder533 replied:
Good catch.

roshandathathri commented on lonestar/analytics/cpu/bipart/Bipart.cpp:
Not a good idea to use std allocators in a parallel section. Use gstl::Set instead. It might be better to also use thread-local storage so that the allocation is reused for different nodes executed by the same thread (you'll need to clear the set after each node is executed).

roshandathathri commented on lonestar/analytics/cpu/bipart/Bipart.cpp:
It seems a bit weird to use a set when all you care about the set is that it has more than 1 element. 

For this use case, it is sufficient to have a single variable initialized to the partition of the first edge
and a boolean to check whether any edge had a different partition or not. If any edge has a different partition, you can break about of the for loop.

roshandathathri commented on lonestar/analytics/cpu/bipart/Bipart.cpp:
Are these vectors expected to allocated already?

Also if these are read-write parameters, Katana coding convention is to use pointers instead of references.

nicelhc13 replied:
Yes it expects that the vectors are already allocated.

roshandathathri commented on lonestar/analytics/cpu/bipart/Bipart.cpp:
Better suited in a separate cpp file.

roshandathathri commented on lonestar/analytics/cpu/bipart/Bipart.cpp:
The difference between Partition and CreatePartitions is not clear in their names or doxygen documentation.

roshandathathri commented on lonestar/analytics/cpu/bipart/Bipart.cpp:
CreatePartitions() is too long and would benefit from having more sub-functions.

roshandathathri commented on lonestar/analytics/cpu/bipart/Coarsening.cpp:
```diff
@@ -17,560 +17,1088 @@
+
+        GNode hedge           = hedge_index_pair.first;
+        MetisNode& hedge_data = fine_graph->getData(hedge);
+        std::vector<GNode> edges;
```
Same comment as std::set. Use gstl::Vector instead. This might also benefit from using thread-local storage in addition to gstl::Vector. 

Check here for an example:
https://github.com/KatanaGraph/katana/blob/master/lonestar/analytics/cpu/gmetis/Coarsening.cpp#L175

roshandathathri commented on lonestar/analytics/cpu/bipart/Coarsening.cpp:
```diff
@@ -17,560 +17,1088 @@
+        MetisNode& hedge_node = fine_graph->getData(hedge);
+        WeightTy best_weight{max_weight};
+        GNode best_node{0};
+        VecTy cells;
```
Same comment as std::vector.

roshandathathri commented on lonestar/analytics/cpu/bipart/Coarsening.cpp:
```diff
@@ -17,560 +17,1088 @@
+        uint32_t h_index      = hedge_index_pair.second;
+        GNode hedge           = hedge_index_pair.first;
+        if (hedges[h_index].test(hedge)) {
+          num_wip_hg[h_index] += 1;
```
This can suffer from cache conflicts. A better design would to be per-thread storage of vectors (without using GAccumulator).

roshandathathri commented on lonestar/analytics/cpu/bipart/Coarsening.cpp:
```diff
@@ -17,560 +17,1088 @@
+          num_wip_hg[h_index] += 1;
+        }
+      },
+      galois::steal(), galois::loopname("Coarsening-Count-HEdges"));
```
Stealing might hurt the performance in this case. Unless the load for each operator varies, the overheads of stealing outweigh its benefits.

Typically, use steal only if edges are being iterated over within an operator.

roshandathathri commented on lonestar/analytics/cpu/bipart/Coarsening.cpp:
```diff
@@ -17,560 +17,1088 @@
-    if (hedges[n]) {
-      old_id[h_id]                  = fineGGraph->getData(n).netnum;
-      fineGGraph->getData(n).nodeid = h_id++;
+      galois::steal(), galois::loopname("Coarsening-Count-PostponedNodes"));
```
Ditto

nicelhc13 replied:
It iterates edges

roshandathathri commented on lonestar/analytics/cpu/bipart/Coarsening.cpp:
```diff
@@ -17,560 +17,1088 @@
+  std::vector<std::vector<uint32_t>> idmap(num_partitions);
+  std::vector<std::vector<WeightTy>> new_weight(num_partitions);
+
+  // TODO(HCLee) I don't know why it could not be parallelized.
```
Drop the comment and the commented code

@nicelhc13 GAccumulator reduce() is NOT valid within a parallel region.
https://github.com/KatanaGraph/katana/blob/master/libgalois/include/galois/Reduction.h#L100
This might be why you were not able to parallelize this. Anyway, number of partitions is expected to be low and the operator for each partition has very little compute, so I don't think it's worth parallelizing this.

roshandathathri commented on lonestar/analytics/cpu/bipart/Coarsening.cpp:
```diff
@@ -17,560 +17,1088 @@
+        }
+        uint32_t num_hedges = fine_graphs[i]->hedges;
+        uint32_t tot_size   = fine_graphs[i]->size();
+        galois::DynamicBitSet new_match_filter;
```
Better to use PerThreadStorage and clear()

roshandathathri reviewed (commented):
I didn't look at Coarsening.cpp in detail and I didn't look at Partitioning.cpp, and Refine.cpp.

Some high-level comments:
- Please have a clear policy to choose one among these data structures for an array: LargeArray, PODResizeableArray, gstl::Vector, std::vector. If the policy is not clear, we can discuss on slack. 
- Please use more PerThreadStorage when an array (or similar) variable within an operator can allocate a large chunk of memory.
- Please avoid using galois::steal unless the operator is iterating over edges or is expected to have a highly varied amount of computation.

Some of my inline comments in nits. I'll leave it to you to decide whether to handle them or not.

nicelhc13 commented on libgalois/include/galois/graphs/LC_CSR_Graph.h:
```diff
@@ -760,6 +760,65 @@ class LC_CSR_Graph :
+
+    edgeIndData = std::move(prefix_sum);
+
+    galois::do_all(galois::iterate((uint32_t)0, numNodes), [&](uint32_t n) {
```
(uint32_t)0 --> uint32_t{0}

nicelhc13 commented on libgalois/include/galois/graphs/LC_CSR_Graph.h:
```diff
@@ -760,6 +760,65 @@ class LC_CSR_Graph :
+
+    edgeIndData = std::move(prefix_sum);
+
+    galois::do_all(galois::iterate((uint32_t)0, numNodes), [&](uint32_t n) {
```
Same as the above

nicelhc13 commented on lonestar/analytics/cpu/bipart/Bipart.h:
Removing GCopyableAtomic is fine but I think it would be better to keep 44~49 lines.
They are used in other files.

nicelhc13 commented on lonestar/analytics/cpu/bipart/Coarsening.h:
From the Roshan's comment, we should modify this to pointer types since we are modifying them,

nicelhc13 commented on lonestar/analytics/cpu/bipart/Coarsening.h:
I kept thinking if renaming this function would be possible. First, it is asymmetric. we don't have CoarsePhaseI(). Second, we cannot expect which algorithm is applied to here with this name. but i have no strong preference. 
what do you think @uditagar  ?

uditagar replied:
CoarseUnmatchedNodes maybe, since this phase deals with corasening the leftover nodes

nicelhc13 commented on lonestar/analytics/cpu/bipart/Coarsening.h:
we would need to update them to pointer types.

nicelhc13 commented on lonestar/analytics/cpu/bipart/Coarsening.h:
@uditagar  do you have a good name for this?

nicelhc13 commented on lonestar/analytics/cpu/bipart/Coarsening.h:
@uditagar I still don't understand why it cannot be parallelized. Does it because this semantic requires hard ordering? When I changed this loop to do_all(), sanity check failed.

uditagar replied:
Roshan mentioned that reduce is parallel too, that's why it failed. My suggestion is to put hnum[i] in a for loop and the rest in do_all

nicelhc13 commented on lonestar/analytics/cpu/bipart/Helper.h:
```diff
@@ -0,0 +1,440 @@
+        g.getData(n).SetPositiveGain(gain);
+      },
+      galois::steal(), galois::loopname("Reduce-Gains"));
+}
```
@uditagar  do you want to keep using this implementation? or do you want to replace it with your new version using modulo function?

uditagar replied:
the modulo didn't work either and this is also slow. The current version that I have is to have these as global vectors, that is fast. I will update it once Gill approves it.

gurbinder533 replied:
Let me merge this PR and then we can make more changes.

gurbinder533 commented:
I think this PR is in a reasonable state to merge. There are a few more changes that are required, but we can do that in subsequent PRs and share the work.

roshandathathri commented on lonestar/analytics/cpu/bipart/Partitioning.h:
Please move the definition of these functions to cpp files.

roshandathathri commented on lonestar/analytics/cpu/bipart/Refine.h:
Ditto.

roshandathathri commented on lonestar/analytics/cpu/bipart/Helper.h:
```diff
@@ -0,0 +1,434 @@
+ * @param graph Graph
+ * @param num_hedges Number of hyperedges in the specified param graph
+ */
+void InitNodes(GGraph& graph, uint32_t num_hedges) {
```
Ditto.

roshandathathri commented on lonestar/analytics/cpu/bipart/Coarsening.h:
Ditto.

roshandathathri reviewed (approved):
> I think this PR is in a reasonable state to merge. There are a few more changes that are required, but we can do that in subsequent PRs and share the work.

I'm fine with this. Please keep track of all unresolved comments in this PR so that you can revisit them later.

My main new comment:

- Do not include definitions of non-templated functions (that have more than a couple of lines) in the header. Move them to a cpp file.

roshandathathri commented on lonestar/analytics/cpu/bipart/Helper.h:
```diff
@@ -0,0 +1,43 @@
+ * @param prefix_sum A LargeArray to store prefix sum of nodes
+ */
+template <typename T>
+uint64_t ParallelPrefixSum(galois::LargeArray<T>& prefix_sum);
```
Template function need to be defined in the header. Otherwise, they can't be used in any file that does not have its definition. 

If they are only used in Helper.cpp, then drop this declaration.

roshandathathri reviewed (approved):
LGTM

Commit:
PropertyGraph: move non-typical access functions to another class

Sometimes we have to to do messy things to PropertyGraphs especially
when partitioning them. In an effort to keep those interfaces from
confusing direct users of PropertyGraph, this PR factors them out into
a friend class that I'm calling a "Retractor". The retractor lets
other codes that knows what it's doing have controlled access to
PropertyGraph internals.

This was motivated by the smaller change which added methods to
deallocate topology information. (Required to save memory during
repartitioning).

Signed-off-by: Tyler Hunt <thunt@katanagraph.com>

tylershunt commented:
Paired with https://github.com/KatanaGraph/katana-enterprise/pull/1784

amberhassaan commented on libtsuba/include/tsuba/RDG.h:
```diff
@@ -229,6 +229,9 @@ class KATANA_EXPORT RDG {
   void DropEdgeProperties();
 
+  /// Remove topology data
+  katana::Result<void> DropToplogy();
```
This method has a typo in the name everywhere

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -95,57 +95,7 @@ class KATANA_EXPORT PropertyGraph {
-  const std::shared_ptr<arrow::Table>& edge_properties() const {
-    return rdg_.edge_properties();
-  }
+  friend class PropertyGraphRetractor;
```
I'm going to insist on a better name. More importantly, we should define the purpose of this class clearly. 

A less urgent but important issue at hand is that PropertyGraph should be made unmoveable so that all parties keep a unique_ptr<PropertyGraph> or shared_ptr<> if that makes sense in certain scenarios.

tylershunt replied:
The purpose of this class is to allow:
 * access to data that is only useful when building/using partitioned graphs
 * access to methods that will leave PropertyGraph in a state where it's other invariants don't hold.

Do you have any suggestions for a better name?

tylershunt replied:
`PropertyGraphDistributedMetaDataManager`? `BandaidForBadLeakyAbstraction`? ;)

tylershunt replied:
> A less urgent but important issue at hand is that PropertyGraph should be made unmoveable so that all parties keep a unique_ptr or shared_ptr<> if that makes sense in certain scenarios.

Maybe but I'm not going to address that here.

witchel replied:
I have not reviewed the code yet, but OMG, I thought PropertyGraphRetractor was the most brilliant name I've ever heard.  I loved the TV show MASH when I was a kid, so I know that "nurse, retractor" means its time to get into the guts.  Perfect.

amberhassaan replied:
Something like PropertyGraphInternals or PropertyGraphDetails that hints that it's not for everybody's use.

witchel commented on libgalois/include/katana/PropertyGraphRetractor.h:
```diff
@@ -0,0 +1,129 @@
+  /// Tell the RDG where it's data is coming from
+  Result<void> InformPath(const std::string& input_path);
+
+  // TODO(witchel): ChunkedArray is inherited from arrow::Table interface but this is
+  // really a ChunkedArray of one chunk, change to arrow::Array.
```
This is irrelevant to this PR, but can you give me the ChunkedArray pep talk again. Why can't we turn everything into Arrays?  I thought those could get as large as physical memory.

tylershunt replied:
Normal arrays can be large yea, but forcing us to use them everywhere means we will always need space to unchunk them (things come chunked out of parquet and we need 2x space since we have to copy everything to fix that). I think there's a possible future where that assumption is not tenable. We've already started avoiding unchunking things during import as much as possible.

witchel commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -412,7 +320,7 @@ class KATANA_EXPORT PropertyGraph {
   /// get the schema for loaded node properties
   std::shared_ptr<arrow::Schema> loaded_node_schema() const {
-    return node_properties()->schema();
+    return rdg_.node_properties()->schema();
```
You made the `rdg_` explicit for clarity?

tylershunt replied:
I noticed only calls to `node/edge_properties()`  were now internal to this class (i think in large part thanks to you!), so I thought I would just ditch those pass throughs to reduce clutter.

witchel reviewed (approved):
Sorry, I've started this review several times and then wandered away without actually committing.  I think moving these "bad abstraction" methods out of `PropertyGraph `is a great thing.
